### PR TITLE
Adding allergy type, category, reactions and severity

### PR DESCRIPTION
**Pull request by eedrummer**
Breaking up allergy onsets into separate submodules based on category
Adding drug allergies

### CODE REVIEW

**Comment by eedrummer** [comment]
This PR now also includes reactions with severity levels.
All codes for allergies have been switched over to substance codes. So the codes are now for "substance" as opposed to "allergic to substance" codes.
Exporter updated for FHIR R4, CCDA and CSV to support the new allergy features.

File: src/main/resources/synthea.properties [hunk 1]
```diff
@@ -14,11 +14,11 @@ exporter.fhir.export = true
 exporter.fhir_stu3.export = false
 exporter.fhir_dstu2.export = false
 exporter.fhir.use_shr_extensions = false
-exporter.fhir.use_us_core_ig = true
+exporter.fhir.use_us_core_ig = false
 exporter.fhir.transaction_bundle = true
 exporter.fhir.bulk_data = false
 exporter.groups.fhir.export = false
-exporter.hospital.fhir.export = true
+exporter.hospital.fhir.export = false
```
Comment (jawalonoski) [inline src/main/resources/synthea.properties:21]: I think you want to revert these changes.

US Core should be default, so should the hospital export.

**Review by jawalonoski — CHANGES_REQUESTED** [review]
A few minor adjustments needed, I think.

Also, there are some check style complaints:

```
ReactionProbabilitiesTest.java:8: Wrong lexicographical order for 'java.util.ArrayList' import. Should be before 'org.mitre.synthea.world.concepts.HealthRecord'. [CustomImportOrder]
ReactionProbabilitiesTest.java:9: Wrong lexicographical order for 'java.util.HashMap' import. Should be before 'org.mitre.synthea.world.concepts.HealthRecord'. [CustomImportOrder]
ReactionProbabilitiesTest.java:10: Wrong lexicographical order for 'java.util.List' import. Should be before 'org.mitre.synthea.world.concepts.HealthRecord'. [CustomImportOrder]
ReactionProbabilitiesTest.java:12: Import statement for 'org.junit.Assert.*' is in the wrong order. Should be in the 'STATIC' group, expecting not assigned imports on this line. [CustomImportOrder]
ReactionProbabilitiesTest.java:12: Using the '.*' form of import should be avoided - org.junit.Assert.*. [AvoidStarImport]
ReactionProbabilitiesTest.java:17:3: Missing a Javadoc comment. [JavadocMethod]
ReactionProbabilitiesTest.java:21: Line is longer than 100 characters (found 113). [LineLength]
ReactionProbabilitiesTest.java:22: Line is longer than 100 characters (found 113). [LineLength]
ReactionProbabilitiesTest.java:43: Line is longer than 100 characters (found 117). [LineLength]
ReactionProbabilitiesTest.java:44: Line is longer than 100 characters (found 113). [LineLength]
ReactionProbabilitiesTest.java:45: Line is longer than 100 characters (found 113). [LineLength]
StateTest.java:967:84: ')' is preceded with whitespace. [ParenPad]
StateTest.java:969:67: '||' should be on a new line. [OperatorWrap]
```

File: src/main/java/org/mitre/synthea/engine/Logic.java [hunk 2]
```diff
@@ -422,49 +422,96 @@ public boolean test(Person person, long time) {
...
+        if (person.attributes.containsKey(this.referencedByAttribute)) {
+          return checkAttribute(person,
+              (HealthRecord.Entry) person.attributes.get(this.referencedByAttribute));
         } else {
           return false;
         }
       }
 
-      throw new RuntimeException("Active Condition logic must be specified by code or attribute");
+      throw new RuntimeException("Active CarePlan logic must be specified by code or attribute");
```
[the remark this answers is not in the document]
Comment (dehall) [reply src/main/java/org/mitre/synthea/engine/Logic.java:461]: If we want it to be more specific we could use `this.getClass().getSimpleName()`

File: src/main/java/org/mitre/synthea/engine/ReactionProbabilities.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+import org.apache.commons.math3.distribution.EnumeratedDistribution;
+import org.apache.commons.math3.util.Pair;
+import org.mitre.synthea.world.agents.Person;
+import org.mitre.synthea.world.concepts.HealthRecord;
+
+/**
+ * ReactionProbabilities is a structure to capture the kind and severity of reaction for a
+ * given AllergyIntolerance.
+ */
+public class ReactionProbabilities implements Serializable {
```
Comment (dehall) [inline src/main/java/org/mitre/synthea/engine/ReactionProbabilities.java:16]: Just a note that I was initially a little confused by jumping back & forth between reaction "severities" used in some places and reaction "probabilities" used here. Would be nice to have those a little more consistent but not a big deal.

File: src/main/java/org/mitre/synthea/export/CSVExporter.java [hunk 4]
```diff
@@ -724,7 +728,40 @@ private void allergy(String personID, String encounterID, Entry allergy) throws
...
 
     s.append(coding.code).append(',');
-    s.append(clean(coding.display));
+    s.append(clean(coding.display)).append(',');
+
+    int reactionsSize = 0;
+    if (allergy.reactions != null) {
+      reactionsSize = allergy.reactions.size();
+    }
+    Function<Map.Entry<HealthRecord.Code, HealthRecord.ReactionSeverity>, String> template =
```
Comment (dehall) [inline src/main/java/org/mitre/synthea/export/CSVExporter.java:737]: Ain't Java beautiful?

File: src/main/java/org/mitre/synthea/export/CSVExporter.java [hunk 5]
```diff
@@ -724,7 +728,40 @@ private void allergy(String personID, String encounterID, Entry allergy) throws
...
+    switch (reactionsSize) {
+      case 0:
+        s.append(",,,");
+        break;
+      case 1:
+        s.append(allergy.reactions.entrySet().stream().map(template).collect(Collectors.joining()));
+        s.append(",,");
+        break;
+      case 2:
+        s.append(allergy.reactions.entrySet().stream().map(template)
```
Comment (dehall) [inline src/main/java/org/mitre/synthea/export/CSVExporter.java:754]: We could remove this case and fall back to default if there are 2 reactions, to ensure that reaction1 is always the more severe of the 2. Not sure if that matters.

File: src/main/java/org/mitre/synthea/export/Exporter.java [hunk 6]
```diff
@@ -491,7 +491,14 @@ private static HealthRecord filterForExport(HealthRecord record, int yearsToKeep
       filterEntries(encounter.conditions, claimItems, cutoffDate, endTime, keepCondition);
 
       // allergies are essentially the same as conditions
-      filterEntries(encounter.allergies, claimItems, cutoffDate, endTime, keepCondition);
+      // But we need to redefine all of the predicates, because we are talking about Allergies as
+      // opposed to Entries... You would think that it would work... but generics are hard
+      Predicate<HealthRecord.Allergy> allergyActive = c -> record.conditionActive(c.type);
```
Comment (dehall) [inline src/main/java/org/mitre/synthea/export/Exporter.java:496]: It does just call out to `conditionActive` behind the scenes, but let's use `record.allergyActive` here
```suggestion
      Predicate<HealthRecord.Allergy> allergyActive = c -> record.allergyActive(c.type);
```

File: src/main/java/org/mitre/synthea/export/CSVExporter.java [hunk 7]
```diff
@@ -709,8 +712,9 @@ private void condition(String personID, String encounterID, Entry condition) thr
    * @param allergy     The allergy itself
    * @throws IOException if any IO error occurs
    */
-  private void allergy(String personID, String encounterID, Entry allergy) throws IOException {
-    // START,STOP,PATIENT,ENCOUNTER,CODE,DESCRIPTION
+  private void allergy(String personID, String encounterID, HealthRecord.Allergy allergy)
+      throws IOException {
+    // START,STOP,PATIENT,ENCOUNTER,CODE,DESCRIPTION,REACTION1,SEVERITY1,REACTION2,SEVERITY2
```
Comment (dehall) [inline src/main/java/org/mitre/synthea/export/CSVExporter.java:717]: Looking at the output here, I'm wondering if we want to add a couple more columns.
Sample: [allergies.csv](https://github.com/synthetichealth/synthea/files/6678076/allergies.csv)

- Either code system or type would be useful here, to know whether the allergy is to a medication or something else. A person can look at it and know penicillin is a medication so it's an RxNorm code, pet dander is a SNOMED code, but for anything automated it would be nice to be explicit.
- The reactions at the end are just codes with no description. I'd be fine with keeping just the code, since it's consistently SNOMED and can be looked up, but just about everywhere else in the CSV exporter we include the description with the code. (And there's at least one place we use the description without the code)

**Review by dehall — APPROVED** [review]
I have some suggestions but nothing is show-stopping. @jawalonoski feel free to merge when you're ready

File: src/main/java/org/mitre/synthea/export/FhirR4.java [hunk 8]
```diff
@@ -1473,6 +1497,29 @@ private static BundleEntryComponent allergy(RandomNumberGenerator rand,
     Code code = allergy.codes.get(0);
     allergyResource.setCode(mapCodeToCodeableConcept(code, SNOMED_URI));
```
Comment (jawalonoski) [inline src/main/java/org/mitre/synthea/export/FhirR4.java:1498]: We can assume `SNOMED_URI` should be used here, since a drug allergy/intolerance could instead require using `RXNORM_URI`. Some more logic is required here.

Comment (dehall) [reply src/main/java/org/mitre/synthea/export/FhirR4.java:1498]: I think this is only used as a default in case the provided `code` doesn't contain a system for some reason. I was looking at some FHIR resources and the actual URI used was RxNorm when appropriate, ex:
```
{
      "resourceType": "AllergyIntolerance",
      "id": "4214370a-a980-d3a6-ecb2-d601c1423848",
      "clinicalStatus": {
        "coding": [ {
          "system": "http://terminology.hl7.org/CodeSystem/allergyintolerance-clinical",
          "code": "active"
        } ]
      },
      "verificationStatus": {
        "coding": [ {
          "system": "http://terminology.hl7.org/CodeSystem/allergyintolerance-verification",
          "code": "confirmed"
        } ]
      },
      "type": "allergy",
      "category": [ "medication" ],
      "criticality": "low",
      "code": {
        "coding": [ {
          "system": "http://www.nlm.nih.gov/research/umls/rxnorm",
          "code": "7984",
          "display": "Penicillin V"
        } ],
        "text": "Penicillin V"
      },
      "patient": {
        "reference": "urn:uuid:e27548c2-8384-8208-17dc-aeb0587bdecc"
      },
      "recordedDate": "1997-12-15T03:30:39-05:00",
      "reaction": [ {
        "manifestation": [ {
          "coding": [ {
            "system": "http://snomed.info/sct",
            "code": "247472004",
            "display": "Wheal (finding)"
          } ],
          "text": "Wheal (finding)"
        } ],
        "severity": "mild"
      }, {
        "manifestation": [ {
          "coding": [ {
            "system": "http://snomed.info/sct",
            "code": "271807003",
            "display": "Eruption of skin (disorder)"
          } ],
          "text": "Eruption of skin (disorder)"
        } ],
        "severity": "mild"
      } ]
    },
```

Comment (jawalonoski) [reply src/main/java/org/mitre/synthea/export/FhirR4.java:1498]: Ah, you are correct. `mapCodeToCodeableConcept` only uses that system by default if the code doesn't already have one.

File: src/main/java/org/mitre/synthea/engine/ReactionProbabilities.java [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+  private static final String SEVERE = "severe";
```
[the remark this answers is not in the document]
Comment (eedrummer) [reply src/main/java/org/mitre/synthea/engine/ReactionProbabilities.java:17]: It could be. For some reason, I thought this would be a better solution than having two separate enums, as I define an enum in HealthRecord that maps the severity to SNOMED codes. Not sure I can justify it.

**Review by jawalonoski — APPROVED** [review]
